dot/parachain/grid/grid_topology.go

```diff
@@ -0,0 +1,352 @@
+	if len(gt.ShuffledIndices) != len(gt.CanonicalShuffling) {
+		return nil, fmt.Errorf("grid topology malformed: "+
+			"shuffledIndices length %d is not equal to canonicalShuffling length %d",
+			len(gt.ShuffledIndices),
+			len(gt.CanonicalShuffling),
+		)
+	}
+
+	shuffledIndex := gt.ShuffledIndices[vi]
+
+	neighbours, err := calculateMatrixNeighbours(shuffledIndex, uint(len(gt.ShuffledIndices)))
+	if err != nil {
+		return nil, err
+	}
+
+	gridSubset := NewEmptyGridNeighbours()
+
+	for _, rN := range neighbours.RowNeighbours {
+		n := &gt.CanonicalShuffling[rN]
+		gridSubset.ValidatorIndicesRow[n.ValidatorIndex] = struct{}{}
+		for _, p := range n.Peers {
+			gridSubset.PeersRow[p] = struct{}{}
+		}
+	}
+
+	for _, cN := range neighbours.ColumnNeighbours {
+		n := &gt.CanonicalShuffling[cN]
+		gridSubset.ValidatorIndicesCol[n.ValidatorIndex] = struct{}{}
+		for _, p := range n.Peers {
+			gridSubset.PeersCol[p] = struct{}{}
+		}
+	}
+
+	return gridSubset, nil
+}
+
+type GridNeighbours struct {
+	PeersRow map[peer.ID]struct{}
+	PeersCol map[peer.ID]struct{}
+
+	ValidatorIndicesRow map[parachaintypes.ValidatorIndex]struct{}
+	ValidatorIndicesCol map[parachaintypes.ValidatorIndex]struct{}
+}
+
+func NewEmptyGridNeighbours() *GridNeighbours {
+	return &GridNeighbours{
+		PeersRow:            make(map[peer.ID]struct{}),
+		PeersCol:            make(map[peer.ID]struct{}),
+		ValidatorIndicesRow: make(map[parachaintypes.ValidatorIndex]struct{}),
+		ValidatorIndicesCol: make(map[parachaintypes.ValidatorIndex]struct{}),
+	}
+}
+
+// RequiredRoutingByIndex Given the originator of a message as a validator index, indicates the part of the topology
+func (gn *GridNeighbours) RequiredRoutingByIndex(origin parachaintypes.ValidatorIndex, local bool) RequiredRouting {
+	if local {
+		return RequiredRoutingGridXY
+	}
+	_, x := gn.ValidatorIndicesRow[origin]
+	_, y := gn.ValidatorIndicesCol[origin]
+
+	if x && y {
+		// If all works correctly origin peer can't be in both rows and columns. But we leave it anyways
+		return RequiredRoutingGridXY
+	}
+	if !(x || y) {
+		return RequiredRoutingNone
+	}
+	if x && !y {
+		return RequiredRoutingGridY
+	}
+	return RequiredRoutingGridX
+}
+
+// ShouldRouteToPeer indicates does peer should receive a message based on GridTopology and Routing strategy
+func (gn *GridNeighbours) ShouldRouteToPeer(routing RequiredRouting, peer peer.ID) bool {
+	switch routing {
+	case RequiredRoutingAll:
+		return true
+	case RequiredRoutingNone, PendingTopology:
+		return false
+	case RequiredRoutingGridXY:
+		_, x := gn.PeersRow[peer]
+		_, y := gn.PeersCol[peer]
+		return x || y
+	case RequiredRoutingGridX:
+		_, x := gn.PeersRow[peer]
+		return x
+	case RequiredRoutingGridY:
+		_, y := gn.PeersCol[peer]
+		return y
+	default:
+		// No way we get here
+		panic(fmt.Errorf("unknown routing strategy for grid topology: %d", routing))
```

Comment:
`fmt.Errorf` is meant for returning errors, not for use inside panic. Since panic accepts a string directly, it's better to use `fmt.Sprintf` instead.
```suggestion
		panic(fmt.Sprintf("unknown routing strategy for grid topology: %d", routing))
```

This avoids creating an unnecessary error object